report/sections/Evaluation/Benchmark.tex

```diff
@@ -0,0 +1,196 @@
+public static class TaskBuilder
+{
+  public static IEnumerable<Task> BuildClientTaskList<T>(int count, Action<T> action)
+      where T : CodeRunnerClient, new()
+  {
+    List<Action> clientActions = new(count);
+    List<Task> tasks = new(count);
+    
+    for (int i = 0; i < count; i++)
+    {
+        T client = new();
+        clientActions.Add(() => action.Invoke(client));
+    }
+
+    foreach (Action clientAction in clientActions)
+    {
+        tasks.Add(Task.Run(clientAction));
+    }
+    return tasks;
+  }
+}
+\end{lstlisting}
+
+\subsubsection{Baseline benchmark implementation}
+The baseline benchmarks seen in listing \ref{lst:baselineBench} is executed multiple times for all combinations of code submissions and number of clients.
+The benchmark itself measures the time it takes for all clients to post the code submission and receive an HTTP response with status code 200.
+If just one of the submissions cannot be processed an exception is thrown, and the benchmark stops.
+It utilizes the BuildClientTaskList method seen in listing \ref{lst:baselineBench}.
+\begin{lstlisting}[language=CSharp, escapechar=~, caption={C\# code showing the implementation of the benchmark for the baseline}, label={lst:baselineBench}]
+// Attributes excluded
+public class RocketBenchmarks
+{
+    [Params(10, 20, 50, 100)] 
+    public int NumberOfRequests { get; set; }
+    [ParamsSource(nameof(CodeSubmissions))]
+    public CodeSubmission CodeSubmission { get; set; }
+    public static IEnumerable<CodeSubmission> CodeSubmissions => CodeLoader.Load();
+
+    [Benchmark]
+    public void PostAndWaitForResponseReceived()
+    {
+      IEnumerable<Task> clientActions = TaskBuilder.BuildClientTaskList<CodeRunnerClient>(NumberOfRequests,
+        client => client.Post(CodeSubmission)
+        .Result.EnsureSuccessStatusCode());
+        
+      Task.WhenAll(clientActions).Wait();
+    }
+}
+\end{lstlisting}
+
+\subsubsection{Queue system benchmark implementation}
+The benchmark for the queue system seen on listing \ref{lst:queueSystemBench} is similar to the one in listing \ref{lst:baselineBench}, as it utilize the method from listing \ref{lst:TaskBuilder} and the parameterization attributes from BenchmarkDotNet.
+However, the internal implementation of the client varies, as it needs to pull the results of the Test Runner after making an initial request.
+Therefore, an additional parameter for the benchmark is introduced to examine how often a client should try to pull for a result.
+Best case pull time needs to be established to compare it with the baseline component. 
+However, due to the nature of remote communication using the internet it may be necessary to establish which polling times to avoid rather than which gives the best average response times, since the benchmarks does not consider the physical distance between client and server.
```

Comment:
```suggestion
However, due to the nature of online remote communication, it may be necessary to establish which polling times to avoid instead of using the best average response times.
This is because the benchmarks do not consider the physical distance between the client and the server.
```